runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java

```diff
@@ -0,0 +1,213 @@
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.util.state.StateNamespaces;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/** Tests for {@link SimpleDoFnRunner}. */
+@RunWith(JUnit4.class)
+public class SimpleDoFnRunnerTest {
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  @Mock StepContext mockStepContext;
+
+  @Mock TimerInternals mockTimerInternals;
+
+  @Before
+  public void setup() {
+    MockitoAnnotations.initMocks(this);
+    when(mockStepContext.timerInternals()).thenReturn(mockTimerInternals);
+  }
+
+  @Test
+  public void testExceptionsWrappedAsUserCodeException() {
+    ThrowingDoFn fn = new ThrowingDoFn();
+    DoFnRunner<String, String> runner =
+        new SimpleDoFnRunner<>(
+            null,
+            fn,
+            null,
+            null,
+            null,
+            Collections.<TupleTag<?>>emptyList(),
+            mockStepContext,
+            null,
+            WindowingStrategy.of(new GlobalWindows()));
+
+    thrown.expect(UserCodeException.class);
+    thrown.expectCause(is(fn.exceptionToThrow));
+
+    runner.processElement(WindowedValue.valueInGlobalWindow("anyValue"));
+  }
+
+  /**
+   * Tests that a users call to set a timer gets properly dispatched to the timer internals. From
+   * there on, it is the duty of the runner & step context to set it in whatever way is right for
+   * that runner.
+   */
+  @Test
+  public void testTimerSet() {
+    WindowFn<?, ?> windowFn = new GlobalWindows();
+    DoFnWithTimers<GlobalWindow> fn = new DoFnWithTimers(windowFn.windowCoder());
+    DoFnRunner<String, String> runner =
+        new SimpleDoFnRunner<>(
+            null,
+            fn,
+            null,
+            null,
+            null,
+            Collections.<TupleTag<?>>emptyList(),
+            mockStepContext,
+            null,
+            WindowingStrategy.of(new GlobalWindows()));
+
+    // Setting the timer needs the current time, as it is set relative
+    Instant currentTime = new Instant(42);
+    when(mockTimerInternals.currentInputWatermarkTime()).thenReturn(currentTime);
+
+    runner.processElement(WindowedValue.valueInGlobalWindow("anyValue"));
+
+    verify(mockTimerInternals)
+        .setTimer(
+            StateNamespaces.window(new GlobalWindows().windowCoder(), GlobalWindow.INSTANCE),
+            DoFnWithTimers.TIMER_ID,
+            currentTime.plus(DoFnWithTimers.TIMER_OFFSET),
+            TimeDomain.EVENT_TIME);
+  }
+
+  /**
+   * Tests that a users call to set a timer gets properly dispatched to the timer internals. From
```

Comment:
I think this Javadoc is copied from the previous test and doesn't fit this test.